numpy/core/src/multiarray/compiled_base.c

```diff
@@ -500,6 +506,37 @@ binary_search(double key, double arr [], npy_intp len)
     if (key > arr[len - 1]) {
         return len;
     }
+    else if (key < arr[0]) {
+        return -1;
+    }
+
+    if (guess < 0)  {
+        guess = 0;
+    }
+    else if (guess >= len - 1) {
+        guess = len - 2;
```

Comment:
nice catch, swapping the if's and making the else if an if should work too, I'll do some testing with avoiding range checks in an followup